tensorflow/python/data/ops/dataset_ops.py

```diff
@@ -1604,6 +1604,7 @@ class PrefetchDataset(Dataset):
 
   def __init__(self, input_dataset, buffer_size):
     """See `Dataset.prefetch()` for details."""
+    assert buffer_size > 0, "The buffer_size ({}) has to be > 0.".format(buffer_size)
```

Comment:
Instead of this assertion, can you add the following line to the C++ implementation [here](https://github.com/tensorflow/tensorflow/blob/047d7965d2877d7b55f4cdb0d0abdcd733f266a9/tensorflow/core/kernels/prefetch_dataset_op.cc#L39)?

```c++
OP_REQUIRES(ctx, buffer_size > 0, errors::InvalidArgument("buffer_size must be > 0"));
```

Reply:
The fix makes sense. Do we need a timeout test case for `buffer_size=0` on python side?

Reply:
It's complicated by the fact that `buffer_size` can be a `tf.Tensor`. The `assert` will fail in that case... in fact I think it will trigger a different exception about using a `Tensor` as a boolean.

The most thorough way to do it would involve the following steps:

1. Convert `buffer_size` to a `tf.Tensor` to validate its type etc.
2. Attempt to convert it back using `tensor_util.constant_value()`.
3. If the `constant_value()` is not `None`, compare it to `0`.

Reply:
I mean your fix at c++ side is correct, because `buffer_size` is converted to an int64 there. Moreover, do we need to add a test?

Reply:
Ahh, yes, that would be a good idea!

Reply:
So we'd better create a`tensorflow/tensorflow/python/kernel_tests/prefetch_dataset_op_test.py` and add a test case, right?

[range_dataset_op_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/range_dataset_op_test.py) is a good example, and then we can register the python file in the `BUILD` like this:

https://github.com/tensorflow/tensorflow/blob/054b515feec0a3fca4cfb1f29adbf423c9027c3a/tensorflow/python/kernel_tests/BUILD#L2925-L2929

I hope the discussion could be useful for your work, @boeddeker.

Reply:
Thanks for the suggestions. I try to do the modifications on monday. 
But I have installed tensorflow from a whl (I am new to tf), so all my changed will be a guess if the code is correct.

If it is less work for you, I am fine when someone other implements the changes.

Reply:
@boeddeker I understand it. Don't worry, I can take over the issue if you like.

Reply:
@facaiy Thanks, that would be great.